### PR TITLE
Separate api client

### PR DESCRIPTION
**Pull request by mitchellhenke**
Will probably have to sort this out a bit better, as far as which module handles auth headers, etc., but  this is a good start for #11

---
This change is [Reviewable](https://reviewable.io/reviews/getsentry/raven-elixir/20)

### CODE REVIEW

File: lib/sentry.ex [hunk 1]
```diff
@@ -229,13 +207,7 @@ defmodule Sentry do
...
 
-  @spec unix_timestamp :: Number.t
-  defp unix_timestamp do
-    {mega, sec, _micro} = :os.timestamp()
-    mega * (1000000 + sec)
-  end
-
-  @spec unix_timestamp :: String.t
+  @spec iso8601_timestamp :: String.t
   defp iso8601_timestamp do
```
Comment (jeregrine) [inline lib/sentry.ex]: Use the elixir DateTime methods http://elixir-lang.org/docs/stable/elixir/DateTime.html

File: lib/sentry/client.ex [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    unquote(@sentry_client "sentry-elixir/#{Mix.Project.config[:version]}")
+  end
+
+  def request(method, url, headers, body) do
+    case :hackney.request(method, url, headers, body, []) do
+      {:ok, 200, _headers, client} ->
+        case :hackney.body(client) do
+          {:ok, body} ->
+            id = Poison.decode!(body)
+                  |> Dict.get("id")
```
Comment (jeregrine) [inline lib/sentry/client.ex]: Use Map

File: lib/sentry/client.ex [hunk 3]
```diff
@@ -0,0 +1,42 @@
+defmodule Sentry.Client do
+  @sentry_version 5
+  quote do
+    unquote(@sentry_client "sentry-elixir/#{Mix.Project.config[:version]}")
+  end
+
+  def request(method, url, headers, body) do
+    case :hackney.request(method, url, headers, body, []) do
+      {:ok, 200, _headers, client} ->
+        case :hackney.body(client) do
```
Comment (jeregrine) [inline lib/sentry/client.ex]: Break this up into multiple functions or use HTTPotion or something

File: lib/sentry.ex [hunk 4]
```diff
@@ -57,39 +57,24 @@ defmodule Sentry do
...
-    "Sentry sentry_version=#{@sentry_version}, sentry_client=#{@sentry_client}, sentry_timestamp=#{timestamp}, sentry_key=#{public_key}, sentry_secret=#{secret_key}"
-  end
-
   @doc """
   Parses and submits an exception to Sentry if DSN is setup in application env.
   """
   @spec capture_exception(String.t) :: {:ok, String.t} | :error
   def capture_exception(exception) do
+    # TODO: better environment handling
     case Application.get_env(:sentry, :dsn) do
```
Comment (jeregrine) [inline lib/sentry.ex]: Move this to a module attribute, and possibly even do it in the Client.

File: lib/sentry.ex [hunk 5]
```diff
@@ -57,39 +57,24 @@ defmodule Sentry do
   def parse_dsn!(dsn) do
```
Comment (jeregrine) [inline lib/sentry.ex]: Probably parse the DSN in the client.

File: lib/sentry/client.ex [hunk 6]
```diff
@@ -0,0 +1,42 @@
+defmodule Sentry.Client do
+  @sentry_version 5
+  quote do
+    unquote(@sentry_client "sentry-elixir/#{Mix.Project.config[:version]}")
+  end
+
+  def request(method, url, headers, body) do
```
Comment (jeregrine) [inline lib/sentry/client.ex]: Since you can only post I'd probably just do a function that posts the body. I'd also have this function turn it into json.

**Comment by jeregrine** [comment]
Basically I'd move everything the API client cares about into the API client. Examples:
- JSON de/encoding
- DSN parsing and usage

File: lib/sentry.ex (hunk 4 above)
Comment (mitchellhenke) [inline lib/sentry.ex]: this is configurable probably at runtime unfortunately

**Comment by mitchellhenke** [comment]
@jeregrine part of the issue is that it uses the existence of the DSN config to send or not send

File: lib/sentry.ex (hunk 1 above)
Comment (mitchellhenke) [inline lib/sentry.ex]: done

**Pull request closed without merging**